rpcserver.go

```diff
@@ -5243,3 +5256,76 @@ func (r *rpcServer) ChannelAcceptor(stream lnrpc.Lightning_ChannelAcceptorServer
 		}
 	}
 }
+
+// NewMacaroon allows the creation of a new macaroon with custom read and write
+// permissions. No first-party caveats are added since this can be done offline.
+func (r *rpcServer) NewMacaroon(ctx context.Context,
+	req *lnrpc.NewMacaroonRequest) (*lnrpc.NewMacaroonResponse, error) {
+
+	rpcsLog.Debugf("[newmacaroon]")
+
+	var (
+		// TODO(guggero): Refactor into constants that are used for all
+		// permissions in this file. Also expose the list of possible
+		// permissions in an RPC when per RPC permissions are
+		// implemented.
+		validActions  = []string{"read", "write", "generate"}
+		validEntities = []string{
+			"onchain", "offchain", "address", "message",
+			"peers", "info", "invoices", "signer", "macaroon",
+		}
+		stringInSlice = func(a string, slice []string) bool {
+			for _, b := range slice {
+				if b == a {
+					return true
+				}
+			}
+			return false
+		}
+	)
+
+	// Don't allow empty permission list as it doesn't make sense to have
+	// a macaroon that is not allowed to access any RPC.
+	if len(req.Permissions) == 0 {
+		return nil, fmt.Errorf("permission list cannot be empty. "+
+			"specify at least one action/entity pair. supported "+
+			"actions are %v, supported entities are %v",
+			validActions, validEntities)
+	}
+
+	// Validate and map permission struct used by gRPC to the one used by
+	// the bakery.
+	requestedPermissions := make([]bakery.Op, len(req.Permissions))
+	for idx, op := range req.Permissions {
+		if op.Action == "" || !stringInSlice(op.Action, validActions) {
+			return nil, fmt.Errorf("invalid permission action. "+
+				"supported actions are %v", validActions)
+		}
+		if op.Entity == "" || !stringInSlice(op.Entity, validEntities) {
+			return nil, fmt.Errorf("invalid permission entity. "+
+				"supported entities are %v", validEntities)
+		}
```

Comment:
So we now do allow combinations that aren't used? Like peers/generate. Probably ok.

Reply:
Yes, we have no way of dynamically checking this at the moment. This will be addresses in the follow-up PR where the permission lists are refactored.